src/talos/HPOFlagging.py

```diff
@@ -0,0 +1,249 @@
+        genes (set[str]): all genes in this report
+
+    Returns:
+        dict[str, set[str]]: ENSG -> set of HPO ids
+    """
+
+    # lookup of {symbol: ENSG}
+    ensg_map: dict[str, str] = read_json_from_path(ensg2symbol)
+
+    gene_to_phenotype: dict[str, set[str]] = defaultdict(set)
+    with open(g2p_file, encoding='utf-8') as f:
+        for line in f:
+            ncbi_gene_id, gene_symbol, hpo_id, hpo_name, frequency, disease_id = line.split('\t')
+            if gene_symbol in ensg_map and ensg_map[gene_symbol] in genes:
+                gene_to_phenotype.setdefault(ensg_map[gene_symbol], set()).add(hpo_id)
+    return gene_to_phenotype
+
+
+def find_genes_in_these_results(result_object: ResultData) -> set[str]:
+    """
+    identify all the genes we care about finding phenotype matches for
+
+    Args:
+        result_object (ResultData):
+
+    Returns:
+        set[str], all ensembl gene IDs in
+    """
+    ensgs: set[str] = set()
+    for participant in result_object.results.values():
+        for variant in participant.variants:
+            # get the gene ID
+            ensgs.add(variant.var_data.info['gene_id'])
+
+            # for structural variants, add all the LOF'd genes
+            if lof := variant.var_data.info.get('predicted_lof'):
+                ensgs.update(set(lof.split(',')))
+
+    return ensgs
+
+
+def annotate_phenotype_matches(result_object: ResultData, gen_phen: dict[str, set[str]], out_path: str):
+    """
+    for each variant, find any phenotype matches between the participant and gene HPO sets
+
+    checks config for HPOFlagging.strict (bool)
+    if strict, this will test for an exact overlap
+    if strict is False, this will be done semantically via Semsimian
+
+    Args:
+        results (ResultData):
+        gen_phen (dict): mapping of ENSGs to relevant HPO terms
+        out_path (str): path to write results to
+    """
+
+    use_strict = config_retrieve(['HPOFlagging', 'strict'], False)
+
+    min_similarity: float = config_retrieve(['HPOFlagging', 'min_similarity'])
+
+    for participant in result_object.results.values():
+        participant_hpos_dict = {hpo.id: hpo.label for hpo in participant.metadata.phenotypes}
+        participant_hpos = set(participant_hpos_dict.keys())
+
+        for variant in participant.variants:
+            var_gene = variant.var_data.info['gene_id']
+            gene_hpos = gen_phen.get(var_gene, set())
+
+            # under strict matching we require exact overlapping terms
+            if use_strict:
+                hpo_intersection = participant_hpos & gene_hpos
+                if not hpo_intersection:
+                    continue
+                for hpo_id in hpo_intersection:
+                    variant.phenotype_labels.add(f'{hpo_id}: {participant_hpos_dict[hpo_id]}')
+
+            # under standard matching we check for phenotypic similarity
+            elif participant_hpos and gene_hpos:
```

Comment:
I think the logic should be that we **always** run the exact matching then **optionally** also run the similarity matching. As I mentioned, our similarity matching currently has some corner cases where we fail to meet the in IC score even if we have an identical HPO term. This will mitigate that most egregious error at least.